lib/rust/profiling/src/frame_stats.rs

```diff
@@ -0,0 +1,236 @@
+//! This module helps to aggregate various per-frame performance statistics, collected over labeled
+//! time intervals.
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unsafe_code)]
+#![warn(unused_import_braces)]
+
+
+use enso_prelude::*;
+
+use enso_logger::DefaultWarningLogger as Logger;
+use enso_logger::*;
+use serde::Deserialize;
+use serde::Serialize;
+
+
+
+// ====================
+// === IntervalsMap ===
+// ====================
+
+type IntervalsMap = HashMap<String, Bundle>;
+
+thread_local! {
+    static ACTIVE_INTERVALS: RefCell<IntervalsMap> = RefCell::new(IntervalsMap::new());
```

Comment:
What are you thinking about making it `mut static`? Accessing it will require unsafe code, but it would be safe given that we keep it private and thread-local.

Reply:
Just for me to better understand, would your suggestion be for performance reasons or others? I didn't realize that [`thread_local!` is not zero-cost](https://matklad.github.io/2020/10/03/fast-thread-locals-in-rust.html), so it sounds like it could make sense for that reason, I guess? I also don't really know what's the policy on `unsafe` in the project, so I kind of preferred to play it safe initially, but if you suggest I do this, I will! Your question also made me start reading about [unstable `#[thread_local]`](https://github.com/rust-lang/rust/issues/29594); would you have some opinion on this? I would be tempted, but from an initial read, it sounds to me like it's maybe somewhat fickle, so maybe a plain `unsafe` as you suggest is actually still easier to reason about in this case?

Reply:
Initially, I was referring to the use of `RefCell` - its run-time checks slightly affect performance, and here we can safely prove that mutable access to the global static variable is safe (as long as this variable is thread-local). So my suggestion was aimed to mitigate the performance impact of `RefCell` and use something like `thread_local! { static mut ACTIVE_INTERVALS: ... }` instead. 

It turns out, that `std::thread_local` doesn't support mutable statics, which is unfortunate. Now I'm thinking if we need `thread_local` in the first place, and what are the safety requirements for this code. Anyway, having this code is not a big deal, so we can just stop spending time on this and leave it as it is :)

> I also don't really know what's the policy on unsafe in the project

I don't know either! Perhaps it's an oversight in our style guide.

> would you have some opinion on this?

Personally, I'm not a fan of using unstable features, and it's pretty easy to mess up with some bug on a non-standard platform (like WASM). So I don't think it's worth it.

Reply:
> What are you thinking about making it `mut static`? Accessing it will require unsafe code, but it would be safe given that we keep it private and thread-local.

But be wary, that being thread local is not sufficient to make mutable access safe! We must also ensure, that we always keep only one mutable reference to the struct. Having two mutable refs in code is an UB, regardless if it is in a multithreaded environment or not.

And it has justification. Consider iterators: even when you're in a single thread, if we allow multiple mutable references to some vector, then we could modify the structure while we're iterating over it (in my experience, this is quite common pitfall when using c++ iterators).

Perhaps some optimization quirks matter as well.

Of course, in this case we could use unsafe, but we should be extremely careful.

Reply:
> Personally, I'm not a fan of using unstable features, and it's pretty easy to mess up with some bug on a non-standard platform (like WASM). So I don't think it's worth it.

We use some unstable features here and there: there are many useful features which are unstable only because their API is subject to change, but the logic is well tested.

In this case, however, I see a red light in the tracking issue: "Unsound with generators" is not ticked as done, and unsoundness is bad.

Reply:
And the last word: If you are concerned with the performance here, the questions is: how much the performance is affected actually? Because if it is a matter of couple of ifs per frame, I'd say it is not worth the effort.